lib/mailbox-handler.js

```diff
@@ -19,99 +20,92 @@ class MailboxHandler {
-                    {
-                        projection: {
-                            retention: true
-                        }
-                    },
-                    (err, userData) => {
-                        if (err) {
-                            return callback(err);
-                        }
+    async createAsync(user, path, opts) {
+        let mailboxData = await this.database.collection('mailboxes').findOne({ user, path });
 
-                        if (!userData) {
-                            const err = new Error('This user does not exist');
-                            err.code = 'UserNotFound';
-                            err.responseCode = 404;
-                            return callback(err, 'UserNotFound');
-                        }
+        if (mailboxData) {
+            const err = new Error('Mailbox creation failed with code MailboxAlreadyExists');
+            err.code = 'ALREADYEXISTS';
+            err.responseCode = 400;
+            throw err;
+        }
 
-                        mailboxData = {
-                            _id: new ObjectId(),
-                            user,
-                            path,
-                            uidValidity: Math.floor(Date.now() / 1000),
-                            uidNext: 1,
-                            modifyIndex: 0,
-                            subscribed: true,
-                            flags: [],
-                            retention: userData.retention
-                        };
+        const mailboxCountForUser = await this.database.collection('mailboxes').countDocuments({ user });
 
-                        Object.keys(opts || {}).forEach(key => {
-                            if (!['_id', 'user', 'path'].includes(key)) {
-                                mailboxData[key] = opts[key];
-                            }
-                        });
+        if (mailboxCountForUser > (await this.settingsHandler.get('const:max:mailboxes', {}))) {
+            const err = new Error('Mailbox creation failed with code ReachedMailboxCountLimit');
+            err.code = 'OVERQUOTA';
+            err.responseCode = 400;
+            throw err;
+        }
 
-                        this.database.collection('mailboxes').insertOne(mailboxData, { writeConcern: 'majority' }, (err, r) => {
-                            if (err) {
-                                if (err.code === 11000) {
-                                    const err = new Error('Mailbox creation failed with code MailboxAlreadyExists');
-                                    err.code = 'ALREADYEXISTS';
-                                    err.responseCode = 400;
-                                    return callback(err, 'ALREADYEXISTS');
-                                }
-                                return callback(err);
-                            }
+        const userData = await this.database.collection('users').findOne({ _id: user }, { projection: { retention: true } });
```

Comment:
This should probably be the first check - if the user does not exist, there is no need to perform other operations.

Reply:
Made it the first check